src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/NativeLayoutInfoNode.cs

```diff
@@ -0,0 +1,257 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.Diagnostics;
+using System.Collections.Generic;
+
+using Internal.Text;
+using Internal.TypeSystem;
+using Internal.NativeFormat;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    /// <summary>
+    /// Native layout info blob.
+    /// </summary>
+    internal sealed class NativeLayoutInfoNode : ObjectNode, ISymbolNode
+    {
+        private ObjectAndOffsetSymbolNode _endSymbol;
+        private ExternalReferencesTableNode _externalReferences;
+
+        private NativeWriter _writer;
+        private MemoryStream _writerStream;
+
+        private Section _signaturesSection;
+        private Section _ldTokenInfoSection;
+
+        public NativeLayoutInfoNode(ExternalReferencesTableNode externalReferences)
+        {
+            _endSymbol = new ObjectAndOffsetSymbolNode(this, 0, "__nativelayoutinfo_End", true);
+            _externalReferences = externalReferences;
+            _writer = new NativeWriter();
+            _signaturesSection = _writer.NewSection();
+            _ldTokenInfoSection = _writer.NewSection();
+        }
+
+        public ISymbolNode EndSymbol
+        {
+            get
+            {
+                return _endSymbol;
+            }
+        }
+
+        public void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
+        {
+            sb.Append(nameMangler.CompilationUnitPrefix).Append("__nativelayoutinfo");
+        }
+        public int Offset => 0;
+        public override bool IsShareable => false;
+
+        public override ObjectNodeSection Section => ObjectNodeSection.DataSection;
+
+        public override bool StaticDependenciesAreComputed => true;
+
+        protected override string GetName() => this.GetMangledName();
+
+        public void SaveNativeLayoutInfoWriter()
+        {
+            if (_writerStream != null)
+            {
+#if DEBUG
+                // Sanity check... We should not write new items to the native layout after 
```

Comment:
This is a lot of code. Why not just null out the `_writer` so that people will simply AV if they try to generate after we saved? (Instead of hitting a weird assert thousands of lines away from the bug.)